velox/functions/prestosql/aggregates/MergeQDigestAggregate.h

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/exec/Aggregate.h"
+#include "velox/functions/lib/QuantileDigest.h"
+#include "velox/functions/prestosql/types/QDigestType.h"
+#include "velox/vector/DecodedVector.h"
+#include "velox/vector/FlatVector.h"
+
+namespace facebook::velox::aggregate::prestosql {
+
+template <typename T>
+struct QDigestAccumulator {
+  explicit QDigestAccumulator(HashStringAllocator* allocator)
+      : digest_(
+            StlAllocator<T>(allocator),
+            facebook::velox::functions::qdigest::kUninitializedMaxError) {}
+
+  void mergeWith(StringView serialized) {
+    if (serialized.empty()) {
+      return;
+    }
+    digest_.mergeSerialized(serialized.data());
+  }
+
+  int64_t serializedSize() {
+    return digest_.serializedByteSize();
+  }
+
+  void serialize(char* outputBuffer) {
+    // QDigest from qdigest_agg() should always have maxError set
+    VELOX_USER_CHECK_NE(
+        digest_.getMaxError(),
+        facebook::velox::functions::qdigest::kUninitializedMaxError,
+        "QDigest maxError is uninitialized");
+    digest_.serialize(outputBuffer);
+  }
+
+ private:
+  facebook::velox::functions::qdigest::QuantileDigest<T, StlAllocator<T>>
+      digest_;
+};
+
+template <typename T>
+class MergeQDigestAggregate : public exec::Aggregate {
+ public:
+  explicit MergeQDigestAggregate(const TypePtr& resultType)
+      : exec::Aggregate(resultType) {}
+
+  int32_t accumulatorFixedWidthSize() const override {
+    return sizeof(QDigestAccumulator<T>);
+  }
+
+  int32_t accumulatorAlignmentSize() const override {
+    return alignof(QDigestAccumulator<T>);
+  }
+
+  bool accumulatorUsesExternalMemory() const override {
+    return true;
```

Comment:
QDigestAccumulator uses memory from HashStringAllocator that is managed by Velox, so this should be false.